### PR TITLE
Add backwards memory examination for the dereference command

### PR DESCRIPTION
**Pull request by TramadoI**
## Description/Motivation/Screenshots

This commit extends the `dereference` command, allowing to examine memory backwards via a negative `--length` value, analogous to the behaviour of `x`.

## Against which architecture was this tested ?

"Tested" indicates that the PR works *and* the unit test (see `docs/testing.md`) run passes without issue.

 * [ ] x86-32
 * [x] x86-64
 * [ ] ARM
 * [ ] AARCH64
 * [ ] MIPS
 * [ ] POWERPC
 * [ ] SPARC
 * [ ] RISC-V

## Checklist

- [x] My PR was done against the `dev` branch, not `main`.
- [x] My code follows the code style of this project.
- [x] My change includes a change to the documentation, if required.
- [x] If my change adds new code, [adequate tests](docs/testing.md) have been added.
- [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

File: gef.py [hunk 1]
```diff
@@ -8310,13 +8310,23 @@ def do_invoke(self, _: List[str], **kwargs: Any) -> None:
...
         if gef.config["context.grow_stack_down"] is True:
-            from_insnum = nb * (self.repeat_count + 1) - 1
-            to_insnum = self.repeat_count * nb - 1
-            insnum_step = -1
+            if nb > 0:
+                from_insnum = nb * (self.repeat_count + 1) - 1
+                to_insnum = self.repeat_count * nb - 1
+                insnum_step = -1
+            else:
+                from_insnum = 0 + self.repeat_count * nb
```
Comment (Grazfather) [inline gef.py:8318]: Why `0 +`?

File: gef.py [hunk 2]
```diff
@@ -8310,13 +8310,23 @@ def do_invoke(self, _: List[str], **kwargs: Any) -> None:
             return
 
         if gef.config["context.grow_stack_down"] is True:
-            from_insnum = nb * (self.repeat_count + 1) - 1
-            to_insnum = self.repeat_count * nb - 1
-            insnum_step = -1
+            if nb > 0:
+                from_insnum = nb * (self.repeat_count + 1) - 1
+                to_insnum = self.repeat_count * nb - 1
+                insnum_step = -1
```
Comment (Grazfather) [inline gef.py:8316]: we don't have to repeat this line in both cases

File: gef.py [hunk 3]
```diff
@@ -8310,13 +8310,23 @@ def do_invoke(self, _: List[str], **kwargs: Any) -> None:
...
+                to_insnum = nb * (self.repeat_count + 1)
+                insnum_step = -1
         else:
-            from_insnum = 0 + self.repeat_count * nb
-            to_insnum = nb * (self.repeat_count + 1)
-            insnum_step = 1
+            if nb > 0:
+                from_insnum = 0 + self.repeat_count * nb
+                to_insnum = nb * (self.repeat_count + 1)
+                insnum_step = 1
```
Comment (Grazfather) [inline gef.py:8325]: we don't have to repeat this in both cases

File: gef.py [hunk 4]
```diff
@@ -8310,13 +8310,23 @@ def do_invoke(self, _: List[str], **kwargs: Any) -> None:
...
+            else:
+                from_insnum = 0 + self.repeat_count * nb
+                to_insnum = nb * (self.repeat_count + 1)
+                insnum_step = -1
         else:
-            from_insnum = 0 + self.repeat_count * nb
-            to_insnum = nb * (self.repeat_count + 1)
-            insnum_step = 1
+            if nb > 0:
+                from_insnum = 0 + self.repeat_count * nb
```
Comment (Grazfather) [inline gef.py:8323]: why 0+

File: gef.py (hunk 1 above)
Comment (TramadoI) [reply gef.py:8318]: The previous implementation had them, so I simply left them in place. I can't tell why they were placed, so I just left them there in case there's some obsure edge-case that I'm not aware of.

**Review by hugsy — CHANGES_REQUESTED** [review]
Please add tests in `tests/commands/dereference.py`, other than that and graz remarks, it looks good to me

File: tests/commands/dereference.py [hunk 5]
```diff
@@ -28,19 +28,21 @@ def test_cmd_dereference_forwards(self):
 
         cmd = "dereference $sp -l 2"
         setup = [
+            "gef config context.grow_stack_down False",
             "set {char[9]} ($sp+0x8) = { 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x00 }",
             "set {char[9]} ($sp-0x8) = { 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x00 }"
         ]
         res = gdb_start_silent_cmd(cmd=setup, after=cmd)
         self.assertNoException(res)
+        print(res)
```
Comment (Grazfather) [inline tests/commands/dereference.py:37]: remove

File: tests/commands/dereference.py [hunk 6]
```diff
@@ -50,18 +52,20 @@ def test_cmd_dereference_backwards(self):
 
         cmd = "dereference $sp -l -2"
         setup = [
+            "gef config context.grow_stack_down False",
             "set {char[9]} ($sp+0x8) = { 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x41, 0x00 }",
             "set {char[9]} ($sp-0x8) = { 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x00 }"
         ]
         res = gdb_start_silent_cmd(cmd=setup, after=cmd)
         self.assertNoException(res)
+        print(res)
```
Comment (Grazfather) [inline tests/commands/dereference.py:61]: remove

File: gef.py [hunk 7]
```diff
@@ -8310,13 +8310,21 @@ def do_invoke(self, _: List[str], **kwargs: Any) -> None:
...
         else:
-            from_insnum = 0 + self.repeat_count * nb
-            to_insnum = nb * (self.repeat_count + 1)
             insnum_step = 1
+            if nb > 0:
+                from_insnum = self.repeat_count * nb
+                to_insnum = nb * (self.repeat_count + 1)
+            else:
+                from_insnum = nb * (self.repeat_count + 1) + 1
+                to_insnum = (0 + self.repeat_count * nb) + 1
```
Comment (Grazfather) [inline gef.py:8327]: why 0+

File: tests/commands/dereference.py [hunk 8]
```diff
@@ -21,3 +21,51 @@ def test_cmd_dereference(self):
...
+
+        """
+        Assuming the default config of grow_stack_down = False, $sp should look like this:
+        0x00007fffffffd268│-0x0008: "BBBBBBBB"
+        0x00007fffffffd270│+0x0000: 0x0000000000000000   ← $rsp
+        Hence, we want to look at the last line of the output
+        """
+        res = res.splitlines()[-2]
+        self.assertTrue("AAAAAAAA" not in res)
+        self.assertTrue("BBBBBBBB" in res)
```
Comment (Grazfather) [inline tests/commands/dereference.py:71]: assertIn

File: tests/commands/dereference.py [hunk 9]
```diff
@@ -21,3 +21,51 @@ def test_cmd_dereference(self):
...
+        print(res)
+
+        """
+        Assuming the default config of grow_stack_down = False, $sp should look like this:
+        0x00007fffffffd270│+0x0000: 0x0000000000000000   ← $rsp
+        0x00007fffffffd278│+0x0008: "AAAAAAAA"
+        Hence, we want to look at the second last line of the output
+        """
+        res = res.splitlines()[-1]
+        self.assertTrue("AAAAAAAA" in res)
```
Comment (Grazfather) [inline tests/commands/dereference.py:46]: assertIn

File: tests/commands/dereference.py [hunk 10]
```diff
@@ -21,3 +21,51 @@ def test_cmd_dereference(self):
...
+        self.assertTrue("BBBBBBBB" not in res)
```
Comment (Grazfather) [inline tests/commands/dereference.py:47]: assertNotIn

**Review by Grazfather — APPROVED** [review]

**Review by hugsy — APPROVED** [review]
Thanks for the PR!
